plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,203 @@
+
+> in type theory, a term belonging to some type is said to be of canonical form if it is explicitly built up using the constructors of that type. A canonical form is in particular a normal form (one not admitting any reductions), but the converse need not hold.
+>
+> For example, the terms of type `ℕ` (a natural numbers type) of canonical form are the numerals
+`S(S(S(⋯(0)⋯)))`.
+
+Hypothetically, we could allow for handling `VDelay`, `VLamAbs` and `VBuiltin` too by returning `Dunno` when checking, say, two `VLamAbs` for equality instead of error (or `True`/`False` at random), however that would be a very weird builtin to have, which we'll discuss more later.
+
+All in all, given the choice to handle not just SOPs, but also some of the constants, it is proposed to name the equality checking builtin `equalsCanonical` and serialisation builtin `serialiseCanonical`.
+
+## Untyped
+
+Say we didn't have any Typed Plutus Core and only wanted to implement the most straightforward version of `equalsCanonical`, let's see how it could look like.
+
+We'd add the following definitions:
+
+```haskell
+data EqCanonicalResult
+    = EqCanonicalTrue             -- ^ Equal.
+    | EqCanonicalFalse            -- ^ Not equal.
+    | EqCanonicalUnexpectedInput  -- ^ Bad argument such as a lambda or a partially applied builtin.
+
+instance Semigroup EqCanonicalResult where
+    EqCanonicalTrue <> res2 = res2
+    res1            <> _    = res1
+
+instance Monoid EqCanonicalResult where
+    mempty = EqCanonicalTrue
+
+toEqCanonicalResult :: Bool -> EqCanonicalResult
+toEqCanonicalResult True  = EqCanonicalTrue
+toEqCanonicalResult False = EqCanonicalFalse
+
+fromEqCanonicalResult :: EqCanonicalResult -> EvaluationResult Bool
+fromEqCanonicalResult EqCanonicalTrue            = EvaluationSuccess True
+fromEqCanonicalResult EqCanonicalFalse           = EvaluationSuccess False
+fromEqCanonicalResult EqCanonicalUnexpectedInput = EvaluationFailure
+
+class EqCanonical a where
+    eqCanonical :: a -> a -> EqCanonicalResult
+```
+
+to some module, expose `eqCanonical` as a new builtin returning a `Bool` or failing evaluation:
+
+```haskell
+    toBuiltinMeaning _ver EqualsCanonical =
+        makeBuiltinMeaning
+            (\val1 val2 -> fromEqCanonicalResult $ eqCanonical val1 val2)
+            (\_ _ _ -> ExBudgetLast mempty)
+```
+
+and provide support for the whole thing in the CEK machine by implementing `EqCanonical` for `CekValue`:
+
+```haskell
+instance (GEq uni, Closed uni, uni `Everywhere` Eq) => EqCanonical (ArgStack uni fun ann) where
+    eqCanonical EmptyStack = \case
+        EmptyStack -> EqCanonicalTrue
+        _          -> EqCanonicalFalse
+    eqCanonical (ConsStack arg1 args1) = \case
+        ConsStack arg2 args2 -> eqCanonical arg1 arg2 <> eqCanonical args1 args2
+        _                    -> EqCanonicalFalse
+
+instance (GEq uni, Closed uni, uni `Everywhere` Eq) => EqCanonical (CekValue uni fun ann) where
```

Comment:
Also, does this work fine for the polymorphic builtin types too?

Reply:
> This is a pretty strong new constraint on all builtin types!

It's only new in the sense that we used not to have it in the CEK machine, otherwise it existed from the start. 

> Also, does this work fine for the polymorphic builtin types too?

Discovering the magic of builtins? :)

```
>>> :set -XTypeApplications
>>> someValue @_ @DefaultUni [(True,[42 :: Integer])] == someValue [(True,[42 :: Integer])]
True
>>> someValue @_ @DefaultUni [(True,[42 :: Integer])] == someValue True
False
```